apis/policies/v1alpha2/clusteradmissionpolicy_types.go

```diff
@@ -241,3 +241,71 @@ type ClusterAdmissionPolicyList struct {
 func init() {
 	SchemeBuilder.Register(&ClusterAdmissionPolicy{}, &ClusterAdmissionPolicyList{})
 }
+
+func (r *ClusterAdmissionPolicy) SetStatus(status PolicyStatusEnum) {
+	r.Status.PolicyStatus = status
+}
+
+func (r *ClusterAdmissionPolicy) GetPolicyMode() PolicyMode {
+	return r.Spec.Mode
+}
+
+func (r *ClusterAdmissionPolicy) GetModule() string {
+	return r.Spec.Module
+}
+
+func (r *ClusterAdmissionPolicy) IsMutating() bool {
+	return r.Spec.Mutating
+}
+
+func (r *ClusterAdmissionPolicy) GetSettings() runtime.RawExtension {
+	return r.Spec.Settings
+}
+
+func (r *ClusterAdmissionPolicy) GetStatus() *PolicyStatus {
+	return &r.Status
+}
+
+func (r *ClusterAdmissionPolicy) CopyInto(policy *Policy) {
+	*policy = r.DeepCopy()
+}
+
+func (r *ClusterAdmissionPolicy) GetSideEffects() *admissionregistrationv1.SideEffectClass {
+	return r.Spec.SideEffects
+}
+
+func (r *ClusterAdmissionPolicy) GetFailurePolicy() *admissionregistrationv1.FailurePolicyType {
+	return r.Spec.FailurePolicy
+}
+
+func (r *ClusterAdmissionPolicy) GetMatchPolicy() *admissionregistrationv1.MatchPolicyType {
+	return r.Spec.MatchPolicy
+}
+
+func (r *ClusterAdmissionPolicy) GetRules() []admissionregistrationv1.RuleWithOperations {
+	return r.Spec.Rules
+}
+
+func (r *ClusterAdmissionPolicy) GetNamespaceSelector() *metav1.LabelSelector {
+	return r.Spec.NamespaceSelector
+}
+
+func (r *ClusterAdmissionPolicy) GetObjectSelector() *metav1.LabelSelector {
+	return r.Spec.ObjectSelector
+}
+
+func (r *ClusterAdmissionPolicy) GetTimeoutSeconds() *int32 {
+	return r.Spec.TimeoutSeconds
+}
+
+func (r *ClusterAdmissionPolicy) GetObjectMeta() *metav1.ObjectMeta {
+	return &r.ObjectMeta
+}
+
+func (r *ClusterAdmissionPolicy) GetPolicyServer() string {
+	return r.Spec.PolicyServer
+}
+
+func (r *ClusterAdmissionPolicy) GetUniqueName() string {
+	return r.Name
```

Comment:
I know this is a corner case, but there is still room for conflict.

If a cluster admission policy named `namespace-policy` and a policy in namespace `namespace` named `policy` is created, they are going to conflict when reconciling their webhooks.

I think this should do a transformation too, like `clusterwide-${r.Name}` while the namespaced one is `namespaced-${r.Namespace}-${r.Name}`.

Is just an example on strategy and naming, the idea is that we remove the possibility of cluster wide and namespaced policies to ever collide.

Reply:
> I think this should do a transformation too, like clusterwide-${r.Name} while the namespaced one is namespaced-${r.Namespace}-${r.Name}.

Using `clusterwide-${r.Name}` would be a problem for upgrades with existing `ClusterAdmissionPolicies`, as the new code would not find the existing policies. We could create a script to rename all existing `ClusterAdmissionPolicies`. To be honest the name collision is very unlikely to happen, but yeah it could happen.

Maybe just using` namespaced-${r.Namespace}-${r.Name} `or` kubewarden-${r.Namespace}-${r.Name}` for `AdmissionPolicies` ? Collision could still happen, but chances are really low that someone names their namespaces `kubewarden-*` or `namespaced-*`

Reply:
I think there are still chances that a conflict happens, and I think is better to just make it impossible by design, given we have the possibility to do it.

For the "migration" I see two options right now:

1. Documentation. We can specify the steps, mainly:
    1. They run `kubectl delete validating/mutatingwebhookconfiguration` (targeting only the kubewarden ones)
    2. Upgrade, the controller when it starts will create webhooks with the right name.
    3. In this case we have to explain how it's possible to avoid requests from happening while no webhooks are enforced during the "migration".

1. Since this is something that will likely happen in the future, put some safe initialization logic in the kubewarden-controller, this is:
    1. When it starts, it installs a temporary webhook that **will reject absolutely everything** -- except a very well known set of operations that are needed for the migration.
    2. It deletes all the webhooks that belong to kubewarden.
    3. It regenerates the webhooks with the new format.

I think the automatic way is nice, but maybe it's too complex for what we need. I would go ahead with documentation, but given we have the possibility I think we should make cluster-wide names collisions impossible.

Reply:
yes, you are right it's better to make it impossible. I'll update the documentation to specify the steps needed for the migration. I like option 2 for the future :) 

Names changed in commit https://github.com/kubewarden/kubewarden-controller/pull/177/commits/0154a5e40ea2b11c27d17b5ed647b6b41f3edfb6

Reply:
(unresolving, mark as resolved if needed)
Should we open an issue for 1 or 2? I'm bound to forget, and I suppose w shouldn't ship the feature with either 1 or 2.

Reply:
For this release we will add the migration steps to the documentation. It would be great to have an automatic way for the future.
For the migration steps I was thinking of adding them to the change logs.  Do you think we should also add them somewhere else in the documentation?

Reply:
I think the changelogs are not enough. It's good that they are also there, but I think this information belongs primarily to the Release Notes, given this is what users are supposed to look into before upgrading the version.

Changelogs might include release notes as well, but they are more tied to implementation specific and internal details.